### PR TITLE
[cmake] Build python internally for all platforms

### PR DESCRIPTION
**Pull request by fuzzard**
## Description
Allow all platforms to build python as part of the cmake build rather than as a prebuilt lib (ie tools/depends)

## Motivation and context
Unify cpython lib used across all platforms (ie windows)

## How has this been tested?
macos/windows

## What is the effect on users?
N/A

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by matthuisman** [comment]
Cool 😎

File: tools/depends/target/python3/01-win-cmake.patch [hunk 1]
```diff
@@ -0,0 +1,1140 @@
...
++find_package_handle_standard_args(expat CONFIG_MODE)
++find_package_handle_standard_args(openssl CONFIG_MODE)
++find_package_handle_standard_args(bzip2 CONFIG_MODE)
++find_package_handle_standard_args(libffi CONFIG_MODE)
++find_package_handle_standard_args(xz CONFIG_MODE)
++find_package_handle_standard_args(sqlite3 CONFIG_MODE)
++
++# Based on PCbuild\find_python.bat that allows us to cache to NATIVEPREFIX
++add_custom_target(script.find_python
++    COMMAND powershell -Command Invoke-WebRequest https://aka.ms/nugetclidl -OutFile '${NATIVEPREFIX}/bin/nuget.exe'
```
Comment (thexai) [inline tools/depends/target/python3/01-win-cmake.patch:34]:
```suggestion
+    COMMAND @PowerShell "Invoke-WebRequest https://aka.ms/nugetclidl -OutFile '${NATIVEPREFIX}/bin/nuget.exe'"
```

**Review by thexai — COMMENTED** [review]
Quick tested Windows x64 and UWP and no evident runtime issues found but found some build issues when use Debug or build inside Visual Studio IDE:

python3 error when build inside IDE only:
```
2>-- Build files have been written to: V:/kodi-build/build/python3/src/python3-build
2>Performing build step for 'python3'
2>Versión de MSBuild 17.9.8+b34f75857 para .NET Framework
2>
2>  1>Checking Build System
2>  "powershell" no se reconoce como un comando interno o externo,
2>  programa o archivo por lotes ejecutable.
2>C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(254,5): error MSB8066: Custom build for 'V:\kodi-build\build\python3\src\python3-build\CMakeFiles\f2d726eb4375c9b866c9644c42fcec30\script.find_python.rule;V:\kodi-build\build\python3\src\python3\CMakeLists.txt' exited with code 9009. [V:\kodi-build\build\python3\src\python3-build\script.find_python.vcxproj]
2>C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(254,5): error MSB8066: Custom build for 'V:\kodi-build\CMakeFiles\493af69e25b36a4ee45462f18d5c662a\python3-mkdir.rule;V:\kodi-build\CMakeFiles\493af69e25b36a4ee45462f18d5c662a\python3-download.rule;V:\kodi-build\CMakeFiles\493af69e25b36a4ee45462f18d5c662a\python3-update.rule;V:\kodi-build\CMakeFiles\493af69e25b36a4ee45462f18d5c662a\python3-patch.rule;V:\kodi-build\CMakeFiles\493af69e25b36a4ee45462f18d5c662a\python3-configure.rule;V:\kodi-build\CMakeFiles\493af69e25b36a4ee45462f18d5c662a\python3-build.rule;V:\kodi-build\CMakeFiles\493af69e25b36a4ee45462f18d5c662a\python3-install.rule;V:\kodi-build\CMakeFiles\0684bd289fe5fea1a068d3eddc3c67de\python3-complete.rule;V:\kodi-build\CMakeFiles\95f6655570f894385254bb654dbca3bc\python3.rule;V:\kodi\CMakeLists.txt' exited with code 1.
2>Done building project "python3.vcxproj" -- FAILED.
```

pythonmodule-pil not builds in Debug from IDE:
```
1>-- Build files have been written to: V:/kodi-build/build/pythonmodule-pil/src/pythonmodule-pil-build
1>Performing build step for 'pythonmodule-pil'
1>[  1%] Building C object CMakeFiles/_imaging.dir/src/_imaging.c.obj
1>cl : Command line warning D9025: overriding '/Od' with '/O2'
1>cl : Command line error D8016: '/RTC1' and '/O2' command-line options are incompatible
```

python3 build fails (both command line and IDE) when are installed others Windows SDK (e.g.  Windows 8.1 SDK) in host computer. Seems some cmake glitch to targeting dest Windows version here: 

`1>-- Selecting Windows SDK version 10.0.22621.0 to target Windows .`

This message should be usually:

`1>-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.`

When are installed both SDK's (Windows 8.1 and Windows 10) fails with:

```
  Performing configure step for 'python3'
CUSTOMBUILD : CMake warning :  [V:\kodi-build-depends\python3.vcxproj]
  -- Selecting Windows SDK version 8.1 to target Windows .
    Ignoring extra path from command line:

     "V:/kodi/cmake/modules/buildtools"


CUSTOMBUILD : CMake warning :  [V:\kodi-build-depends\python3.vcxproj]
    Ignoring extra path from command line:

     "V:/kodi/project/BuildDependencies/mingwlibs/x64"


  -- The C compiler identification is unknown
  -- The CXX compiler identification is unknown
  CMake Error at CMakeLists.txt:3 (project):
    No CMAKE_C_COMPILER could be found.



  CMake Error at CMakeLists.txt:3 (project):
    No CMAKE_CXX_COMPILER could be found.
```

Seems cmake picks wrong SDK because target Windows version is empty "."

File: tools/depends/target/python3/01-win-cmake.patch (hunk 1 above)
Comment (thexai) [reply tools/depends/target/python3/01-win-cmake.patch:34]: This fix the first issue, the rest I have no idea for now.

Comment (fuzzard) [reply tools/depends/target/python3/01-win-cmake.patch:34]: Greatly appreciated. I'll look into it.

What VS are you using, and is it a specific SDK to install over the "default" for that VS version?
Also, was this for UWP or the Desktop variant (or both)?

Comment (thexai) [reply tools/depends/target/python3/01-win-cmake.patch]: Visual Studio 2022 v17.9.4 (latest)

All is reproducible at build Windows x64

To reproduce "Windows SDKs conflict" you can install standalone Windows 8.1 SDK from here:
https://go.microsoft.com/fwlink/p/?LinkId=323507

and check only install first item:

sdk

**Comment by fuzzard** [comment]
@thexai ive added a commit to propagate some system variables through to the sub build, which fixes the SDK versioning.

```
2>    -- Selecting Windows SDK version 10.0.19041.0 to target Windows 8.1.
```

Appreciate you testing and providing any further feedback.

**Comment by thexai** [comment]
Confirmed Windows SDK's issue is fixed, with both SDK's installed now is choosed the correct one:

`2>-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.`

Also works build python3 target in Debug inside IDE.

Still not works pythonmodule-pil in Debug inside IDE.

```
2>Performing build step for 'pythonmodule-pil'
2>[  1%] Building C object CMakeFiles/_imaging.dir/src/_imaging.c.obj
2>cl : Command line warning D9025: overriding '/Od' with '/O2'
2>cl : Command line error D8016: '/RTC1' and '/O2' command-line options are incompatible
```

**Comment by thexai** [comment]
Another issue:
When build target `build_internal_depends` only from cmake

e.g.:

```
cmake -G "Visual Studio 17 2022" -A x64 -T host=x64 %path_source%

cmake --build . --config "Debug" --target build_internal_depends
cmake --build . --config "Release" --target build_internal_depends
```

Seems only is build `python3` but not others:

build

`pythonmodule-pil` and `pycryptodome` missing. Is intentional?

My workflow is usually build (one time only) depends in Debug and Release in this way after I can use for both BuildSetup.bat
or generate Visual Studio project to work inside IDE in Debug mode.

**Comment by fuzzard** [comment]
Will look into it.

**Comment by thexai** [comment]
Tested latest update and ALL previously issues detected are now solved!  🙂

Desktop x64 builds fine Debug / Release from command line and inside Visual Studio IDE
Also works build only target `build_internal_depends` and after generate VS project and reuse same depends for BuildSetup.bat (Release).

When Jenkins is happy, I think everything will be ready. 
Now it seems that new problems have arisen in UWP but in other dependencies that have nothing to do with it... When the Jenkins things are resolved I will test UWP things again.

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by thexai** [comment]
@fuzzard 

I have investigated build failure in UWP and is real. Seems a side effect of change the correct CMAKE_SYSTEM_NAME in `ModuleHelpers.cmake`

```
      if(DEFINED CMAKE_SYSTEM_NAME)
        list(APPEND CMAKE_ARGS -DCMAKE_SYSTEM_NAME=${CMAKE_SYSTEM_NAME})
      endif()
```

if hack it with:

```
      if(DEFINED CMAKE_SYSTEM_NAME)
        list(APPEND CMAKE_ARGS -DCMAKE_SYSTEM_NAME=Windows)
      endif()
```

Then builds fine but obviously this is not the solution.

The root cause is in `libnfs` CMakeLists.txt that only considers `Windows` name and not `WindowsStore` here:

https://github.com/sahlberg/libnfs/blob/69f9a18748e75df42578c82990a38b724038672b/CMakeLists.txt#L52-L55

Then maybe a proper solution is patch this file with "OR WindowsStore" or more generic `if(MSVC)` condition.

**Comment by fuzzard** [comment]
There's a bunch of failures (flatc, pcre). I'm on holiday at the moment, so I'll look into it further in a week or two

**Comment by fuzzard** [comment]
Alright, i believe UWP is now building correctly.

As mentioned, there were 3 build failures (pcre/libnfs/flatc). The flatc build failure is a clean generic solution.
pcre was just requiring to disabling Additional Security Checks (https://learn.microsoft.com/en-us/cpp/build/reference/sdl-enable-additional-security-checks?view=msvc-170), and nfs required this change as well for one part of its fix.

the last part for libnfs is a god awful hack. As @thexai noted, nfs doesnt recognise anything not "windows" for the CMAKE_SYSTEM_NAME, which obviously wouldnt be the case when we propogate "WindowsStore" down. There are two options, we patch in WindowsStore support. And this may be the better upstreamable option. The other option (ol dirty hack) allows a module to override the CMAKE_SYSTEM_NAME and essentially just send through "Windows" for a uwp build.

I'll look at an actual upstreamable change for WindowsStore, however i want to confirm nfs still works with this change first. Im not at home for a few more days, so i cant test that specifically, so if @thexai has nfs he can test, that would be great as well, but not absolutely necessary. The actual python stuff is the more relevant testing required for this PR.

**Comment by thexai** [comment]
All of the above is solved but a one issue detected:

When build UWP-64 from inside Visual Studio IDE in Debug mode, generated Python DLL names are:

```
2>  -- Install configuration: "Debug"
2>  -- Up-to-date: V:/kodi/project/BuildDependencies/win10-x64/lib/python3.lib
2>  -- Installing: V:/kodi/project/BuildDependencies/win10-x64/bin/python3.dll
2>  -- Up-to-date: V:/kodi/project/BuildDependencies/win10-x64/lib/python311.lib
2>  -- Installing: V:/kodi/project/BuildDependencies/win10-x64/bin/python311.dll
```

But fails LINK to Kodi.exe due expect `python311_d.lib` :

```
23>------ Build started: Project: generate-packaging, Configuration: Debug x64 ------
24>------ Build started: Project: kodi, Configuration: Debug x64 ------
24>LINK : fatal error LNK1104: cannot open file 'python311_d.lib'
24>Done building project "kodi.vcxproj" -- FAILED.
```

I think is correct generate same name `python311.lib` for both Debug and Release in this case. I'm not sure if we should force build RelWithDebInfo always for python3 or only drop _d suffix...

**Comment by thexai** [comment]
The crash in x64 commented in Slack is due this commit:

https://github.com/xbmc/xbmc/pull/24882/commits/c1cbbba8a209337f2529e9810d787500caecb7f3

Reverting changes of this commit, Windows x64 Release/Debug works again.

**Pull request closed without merging**